### PR TITLE
feat(mlx): true grad-norm reporting and compiled global-norm clipping with gradient accumulation

### PR DESCRIPTION
**Pull request by Lyxot**
## Summary

Two changes to the MLX trainer's gradient-norm handling, one commit each:

1. **Report the true gradient norm, behind an opt-in flag when global-norm clipping is off.** For non-global-clip runs, `grad_norm` was previously reconstructed from Adam's second-moment state. That reconstruction is exact only for uninterrupted Adam/AdamW and has several defects: it measures the wrong quantity (the gradient after value/leaf clipping and LoRA+/embedding-LR scaling, not the pre-clip norm global clipping reports); it is wrong on the first step after `resume_from_checkpoint` (its closure-local previous-`v` accumulator restarts at zero while the restored optimizer state does not); it silently reports nothing for Lion (which exposes `betas` but has no `v` state); it skips Adafactor/SGD/Muon with a one-time warning; and it forces an extra host-side evaluation every logging step. Measured impact: under default leaf clipping it understates the true pre-clip norm by 49-69% (median 59%; same-trajectory Gemma3-270M run, per-step aligned); on the first step after a resume it overshoots 2.1x-6.4x depending on accumulated optimizer state; where no clipping binds the two agree to float noise (max 8.7e-7 relative). See the cross-backend section for the CUDA-anchored measurements. This PR deletes the reconstruction and computes the true fp32 global L2 norm in-graph inside the update path.
2. **Enable `mx.compile` for global-norm clipping with gradient accumulation.** The trainer disabled compilation whenever `max_grad_norm > 0` was combined with `gradient_accumulation_steps > 1`. That guard predates the current clipping implementation: `git log -S` shows it landed with the original MLX training path, while the lazy fp32 `_clip_grad_norm_fp32` (pure MLX ops, no host sync) arrived a month later — and clipped single-step training already ran compiled. The guard is removed; compiled clip+accum is validated below. After merging current `main`, this also removes the finite-text shape planner's `compile_ineligible_global_norm` branch (and its pinning test case): #918 added that branch to mirror the old guard, and keeping it would have silently kept single-process clipped accumulation eager. The e2e tests inject a `FiniteTextBatchPlan` so compiled runs stay compile-eligible under #918's shape guard, and the compiled clip+accum bitwise test now exercises the planned path end to end.

## Behavior changes (please read before reviewing)

- **Runs without global-norm clipping no longer display a grad-norm** — console output, `train/grad_norm` in W&B/TensorBoard, and the grad-norm history — unless global-norm clipping is the resolved clip mode or the user sets the new `MLXTrainingConfig(report_grad_norm=True)`. **Standard TRL-driven runs (the notebooks) are unaffected:** TRL's `SFTConfig` defaults `max_grad_norm=1.0`, which forwards to the MLX trainer and keeps them on the global-clip path — verified on the Gemma3-270M notebook settings (byte-identical losses, gradient norms, adapter weights, and log format vs `main`). The absence applies where `max_grad_norm` resolves to 0: direct `MLXTrainingConfig(...)` construction with defaults, or the Unsloth Studio worker, which currently passes 0. The legacy 9-argument step callback keeps its exact signature and receives `None` as its final argument in that case. Rationale: the old fallback value was misleading in the ways listed above, and HF Trainer likewise reports no gradient norm when norm clipping is off. The flag defaults off because the reporting reduction has the same class of peak-memory cost as global clipping itself.
- **Clip+accum users now train compiled.** Loss curves will not bit-match previous eager runs: eager and compiled execution produce different, individually deterministic rounding trajectories in low precision — a pre-existing property visible at multi-billion-parameter scale. `compile=False` reproduces the previous eager behavior exactly.

## Design notes

- `report_grad_norm` is captured as a Python constant before the step closures are built, so each run has exactly one compiled step graph (a runtime condition would add a trace signature per reporting state). With the flag off and no global clipping there is no reporting reduction, scalar output, or evaluation in the graph at all — default leaf/value/no-clip training is numerically and memory-wise unchanged.
- Reporting never changes update numerics. The `grad_accum == 1` fast path keeps its exact update on raw gradients and derives the reported value from a norm-only copy that reproduces the accumulated path's token multiply/divide rounding, so the value is numerically identical to the norm global clipping computes for bf16/fp32 models (fp16 leaves are weighted in fp32 because fp16 cannot represent token counts >= 65520).
- The reported quantity is the norm of the token-normalized gradient: after the accumulation divide, DDP reduction, and LoRA+/embedding-LR ratios; before any clipping and before weight decay — matching what the global-clip path already reported. Under DDP it is computed after the cross-rank reduction, so all ranks report the same value.
- The norm is evaluated in the same `mx.eval` boundary as model/optimizer state (previously a separate early evaluation); `.item()` happens only at the logging boundary.
- The field is appended last in `MLXTrainingConfig`, so every pre-existing field keeps its positional index in the custom `__init__`; the wholesale-config-copy detection treats dumps that predate the field as complete copies, keeping `warmup_ratio` precedence unchanged for older serialized configs.

## Validation

Unit/regression tests (Apple Silicon, real MLX) in `tests/test_mlx_training_e2e_metal.py`: a bitwise update-invariance test (the flag changes no parameter or optimizer-state bit; compiled, bf16, dtypes pinned); an independent-oracle norm test covering value and placement across clip modes, optimizers (including Lion, and SGD with nonzero coupled weight decay to pin that decay is excluded), eager and compiled execution, and accumulation 1-3; a bitwise eager-vs-compiled parity test for clipped accumulation on tiny models; a regression asserting clip+accum compiles at full-step scope without the old disable message; and an injected-failure test pinning that an evaluation-time error after the compiled step propagates rather than triggering a silent eager retry.

Production-scale validation (Qwen/Qwen2.5-VL-3B-Instruct, full BF16, LoRA r=16, chunked-CE loss, sequence lengths cycling 1024/1536/2048 with two 448x448 images per sample, batch 1, accumulation 4, fresh process per run, two repetitions per variant):

- Reporting off: peak allocator memory in default leaf, value-clip, and no-clip modes unchanged vs the pre-change trainer (raw byte ratios at most 1.0000016).
- Reporting on: peak memory within 2.8% of what compiled global clipping itself costs (the documented equal-cost intent); throughput unchanged at 3B (identical steady-state tokens/sec). The cost inverts at small scale: a 270M LoRA run costs no peak memory but about 7% throughput (interleaved repetitions, tight variance) — per-leaf reduction launch overhead against a fast step, another reason the flag defaults off.
- Compiled clip+accum halves peak memory vs eager clipping (12.3 vs 22.5 GiB) and adds zero compile trace signatures after warm-up vs compiled unclipped training (equal trace counts, stable through the measured window).
- Compiled runs are bit-deterministic across fresh processes (losses, reported norms, SHA-256 digests of every parameter and optimizer-state leaf), and strict vs best-effort compile modes are bit-identical.
- Eager-vs-compiled loss divergence for clipped accumulation stays below the untouched unclipped control pair's (max per-step relative delta 0.73x the control's; median 2.25x, within the 3x bound) — the guard removal adds no divergence beyond pre-existing eager/compiled rounding. A simulated corruption (clipping dropped) exceeds the control envelope by 7.8x, so the comparison discriminates real errors.

Notebook parity: the Gemma3-270M notebook's exact settings (public `FastModel` API, 16-bit load, LoRA r=128 on q/k/v/o/gate/up/down, batch 4, `adamw_8bit`, seed 3407, 10 steps on a fixed dataset) produce byte-identical loss history, gradient norms, and adapter-tensor SHA-256 digests between `main` and this branch — and byte-identical again with `report_grad_norm=True` enabled.

```bash
# focused
python -m pytest tests/test_mlx_training_e2e_metal.py -q
# broader selection
python -m pytest tests/ -q -k "mlx and (grad or clip or norm or trainer)"
```

## Cross-backend validation against CUDA

The reported quantity and the deleted value's defects were measured against the CUDA path as an external reference. In global-clip mode — the path the notebooks use — 30 steps of the same Gemma3-270M LoRA configuration (uniform-length data, `max_grad_norm=1.0` on both backends) put the MLX-reported norm within 4.1% of CUDA's HF-reported median grad-norm with 0.996 curve correlation (loss correlation 0.99995).

The two defect measurements used 64 identical dataset rows (verified identical token IDs and batch composition on both backends), removing batch-content and data-order differences so per-step cross-backend comparison is meaningful; backend rounding still compounds across updates, so early steps carry the strongest signal (MLX `adamw`, CUDA the matched `adamw_torch`). CUDA ran `max_grad_norm=1e9` — never binding — so the HF trainer logs its standard pre-clip global norm.

- **Leaf-clip understatement.** MLX trained in its default mode (resolved 1.0 per-leaf L2 clip); CUDA replicated it with an `on_pre_optimizer_step` callback rescaling each parameter gradient to L2 <= 1.0, after HF's norm logging and before the optimizer step (ordering validated in-run to 6.2e-8 max relative error). The branch's norm tracks CUDA's pre-clip norm to a 2.6% median gap over 12 steps (2.3% over steps 1-6); the old reconstruction sits 53.5% below CUDA at the first-six-step median (up to 68.6% on the earliest steps), converging late once few leaves still exceed the threshold. Direct wrong-quantity confirmation: over those early steps the old value matches CUDA's **post**-leaf-clip global norm to a 0.5% median gap — it reported the clipped-gradient scale, while this branch and CUDA report the pre-clip norm.
- **Resume overshoot.** All variants unclipped, 12 steps, checkpoint at step 6, then resumed. Boundary continuity (step-7 norm over the mean of steps 6 and 8): CUDA 1.02, this branch 1.04, old reconstruction 6.43 (62.17 against a ~9.7 neighborhood). The branch's resumed norms exactly equal its uninterrupted baseline at every post-resume step and sit within 1.1% of CUDA at the boundary; CUDA shows no boundary artifact. The spike is confined to reporting — old and new loss histories are bitwise-equal through and after the resume. The 2.1x end of the summary's range came from a small controlled model; here it is 6.4x — the magnitude grows with the second-moment state accumulated before the checkpoint.

Both defect experiments produced bitwise-identical `main`-vs-branch MLX loss histories (update-invariance on real runs, not only in the unit test), and the CUDA reference ran stock HF/TRL semantics — pre-clip norm, no resume artifact — the semantics this PR adopts.

## Known pre-existing test-suite issues (not addressed here)

Combined MLX test runs order-couple a few failures that also occur on `main`: `tests/test_mlx_save_lora_adapters_filter.py` installs the torch-backed simulation shim only when real MLX is not already imported, so its torch mocks can reach the real `mx.save_safetensors` and raise `std::bad_cast`; `tests/test_mlx_adapter_metadata_persistence.py` installs the shim without teardown, which can hijack later MLX imports in real-runtime modules (why the new tests bind their MLX imports at module scope). Fixture repair is left to a follow-up so this PR stays scoped; this branch introduces no new failures relative to `main` (failure sets verified name-identical).

## Coordination

The HF-callbacks rework in #873 touches the same logging loop; whichever lands second should rebase, and the norm-reporting behavior here (computed every optimizer update when active, surfaced at logging boundaries) is the cadence that rebase should preserve. Related user demand for richer MLX training telemetry: unslothai/unsloth#5138.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/trainer.py [hunk 1]
```diff
@@ -2348,6 +2346,28 @@ def _apply_update_direct(grad):
...
+                        # fp16 cannot represent token counts >= 65520: the
+                        # dtype cast becomes inf and zero grads turn nan.
+                        # Weight fp16 leaves in fp32 instead; fp16 reports the
+                        # mathematically exact norm rather than a bit-match of
+                        # the accumulated path's rounding.
+                        return g.astype(mx.float32) * toks_f * inv_toks
+                    return (g * toks_f.astype(g.dtype)) * inv_toks.astype(g.dtype)
+
+                rounded = tree_map(_rounded_for_norm, grad)
+                grad_norm = _global_grad_norm_fp32(rounded)
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/trainer.py]: Left as-is, with reasoning: MLX arrays are lazy, so `tree_map` here builds expression nodes rather than materialized buffers — whether per-leaf intermediates coexist is the scheduler's decision either way, and a fused `tree_reduce` would create the same per-leaf rounded intermediates before summing. We measured this on a 3B BF16 VLM (batch 1, accum 4, seq 1024-2048): this path peaks within 2.8% of the global-clip path, which performs the same cross-tree reduction, so the copy adds no material overhead in practice. The two-step form is kept deliberately because it mirrors the accumulated path's multiply/divide op-for-op, which is exactly what makes the reported value bitwise-equal to the global-clip norm — a property the tests pin.

File: unsloth_zoo/mlx/trainer.py [hunk 2]
```diff
@@ -655,7 +683,14 @@ def __init__(self, *args, **kwargs):
...
         warmup_ratio_default = type(self).warmup_ratio
-        copied_all_fields = len(provided) == len(config_fields)
+        # Fields appended to the dataclass after configs began round-tripping
+        # through full-field dict dumps. A legacy dump that predates them is
+        # still a wholesale copy and must be detected as one, or its copied
+        # default warmup_steps would start overriding a non-default
+        # warmup_ratio.
+        _appended_fields = {"report_grad_norm"}
+        _field_names = {field.name for field in config_fields}
+        copied_all_fields = (_field_names - _appended_fields) <= set(provided)
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/trainer.py]: This became moot in the latest push: the `report_grad_norm` option was removed (the true norm is now always reported in every clip mode), so the appended-field allowlist and the copied-config special-casing no longer exist — the detection is back to the original exact-field-count form.

File: unsloth_zoo/mlx/trainer.py [hunk 3]
```diff
@@ -1127,6 +1162,9 @@ def add_step_callback(self, fn):
 
         fn(step, total_steps, loss, lr, tokens_sec, peak_gb, elapsed,
            num_tokens, grad_norm=None)
+
+        grad_norm is a float only when global-norm clipping is active or
+        ``report_grad_norm=True``; otherwise the argument is passed as None.
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/trainer.py]: Addressed in the latest push: the note is now a `grad_norm:` parameter-style description under the call example, and its content was updated for the current behavior (the value is present on every optimizer step).

File: unsloth_zoo/mlx/trainer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/trainer.py]: Correction to my previous reply: after weighing the reporting cost (the cross-tree reduction in non-global-clip modes), the opt-in design was restored, so this allowlist exists again as pushed. Keeping it as an inline set for now: it is a single-site compatibility shim for exactly one appended field; if a second appended field ever joins it, promoting it to a class-level constant as suggested would be the right move.

File: unsloth_zoo/mlx/trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/trainer.py]: Correction to my previous reply: the opt-in design was restored, so the note again documents the conditional semantics — but it is now formatted as a parameter-style `grad_norm:` line as suggested, rather than trailing prose.

**Comment by Lyxot** [comment]
@codex review

**Comment by Lyxot** [comment]
staging CI pass in https://github.com/Lyxot/unsloth-zoo/pull/4
图片
